### PR TITLE
Add to wallet button in docs (Issue #355)

### LINKED ISSUE #355: "Add to wallet" button the the link token contracts page
We should have an "add to wallet" button on our [LINK token contracts page.](https://docs.chain.link/docs/link-token-contracts/)

This will make it easier for users to add the LINK token to their wallets so they don't have to manually add it themselves. 

I don't have an example off-hand, but I know it's possible.

### PR DESCRIPTION
**Pull request by jamsea**
Fixes #355.
Adds a button to add Chainlink to a user's Ethereum wallet.

image

Here's a link to the standard for adding to wallets that I found in order to implement the code: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-747.md

### CODE REVIEW

**Comment by jamsea** [comment]
I poked around a bit more, using arbitrary HTML in the @readme/markdown plugin doesn't seem to be allowed. It sanitizes all HTML strings by default and strips away all Javascript. There doesn't seem to be a way around this. We need to be able to include the little chunk of JavaScript inside the markdown table to hook into the wallet to add the coin.

The simplest solution would be to convert the whole page to a liquid template and remove markdown entirely for this section. Then adding custom Javascript wouldn't be sanitized. I think for this particular case, that would make the most sense.  The downside, is the tables are converted to normal HTML tables instead of the markdown syntax tables.

Let me know what you think.

File: docs/Developer Reference/link-token-contracts.html [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+| Symbol         | LINK                                                                                                                                     |
+| Decimals       | 18                                                                                                                                       |
+
+{% endmarkdown %}
+
+<script>
+    console.log("hi");
+
+    function addToWallet() {
+        ethereum
```
Comment (jamsea) [inline docs/Developer Reference/link-token-contracts.html:277]: This is where the magic happens

**Comment by PatrickAlphaC** [comment]
@dwightjl

**Comment by dwightjl** [comment]
Sorry I haven't had time to review this. @jamsea Can you rebase and resolve the merge conflicts?

@PatrickAlphaC can you approve this PR too? I'll also go through and test it.

**Comment by PatrickAlphaC** [comment]
@jamsea is this ready to be delivered?

**Comment by jamsea** [comment]
@PatrickAlphaC @dwightjl I have a fuller than usual schedule this week but I can have it wrapped up by Saturday. 👍

**Comment by aelmanaa** [comment]
thanks so much, @jamsea. 

just reworked the original PR as there were a few things to take into account:

- Keep markdown
- Ensure that we are wallet extension is installed (for now, I've coded only support for Metamask. We can support other wallets in the future)
- Add a button "Add to wallet" only near the right contract address (e.g.: if you are connected to Kovan then you'll see a button near the address)
- Reload the page when the user changes the chain

**Comment by aelmanaa** [comment]
@PatrickAlphaC could you have a look please?

**Comment by jamsea** [comment]
@aelmanaa thats a smart solution! I couldn't figure out how to keep the markdown, great job

File: _src/doc-utils.ts [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+  const linkContract = new ethers.Contract(address, LinkToken, provider);
+  let name: string, symbol: string, decimals: BigNumberish;
+  try {
+    name = await linkContract.name();
+    symbol = await linkContract.symbol();
+    decimals = await linkContract.decimals();
+  } catch (error) {
+    throw new Error(`Error occured while trying to fetch linkContract metadata  ${error}`);
+  }
+  /*
```
Comment (obelisk) [inline _src/doc-utils.ts:214]: There should not be commented out checks in production code.

File: _src/doc-utils.ts [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+  console.log(`Chains ${chainId} of params ${JSON.stringify(params)} succesfully added to wallet`);
+};
+
+/**
+ * Functions which validates the format of Link address and interfacs with the contract to make sure
+ * its metadata (e.g.: symbol) is valid
+ * @param address
+ * @param provider
+ */
+const validateLinkAddress = async (address: string, provider: Web3Provider) => {
```
Comment (obelisk) [inline _src/doc-utils.ts:199]: This function just validates an address, not a LINK contract address.

File: _src/doc-utils.ts [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+  // Reload the page if the user changes the Chain.
+  // Only reload the page if the user initiates the switch directly from wallet
+  ethereum.on('chainChanged', (chainId) => {
+    if (chainId !== detectedChainId && chainFromSwitch !== chainId) {
+      window.location.reload();
+    }
+  });
+
+  const tokenAddressElements = Array.from(document.getElementsByClassName('erc-token-address'));
+  const seperator = '_';
```
Comment (obelisk) [inline _src/doc-utils.ts:267]: Nit. Pull to top

File: _src/doc-utils.ts [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+  // Test to follow this pattern: chianId_address (e.g.: 1_0x514910771af9ca656af840dff83e8264ecf986ca)
+  const pattern = new RegExp(`^[0-9]+${seperator}${addressPattern}$`);
```
Comment (obelisk) [inline _src/doc-utils.ts:269]: Nit. Pull to top

File: _src/doc-utils.ts [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+
+  const addToWalletText = 'Add to wallet';
```
Comment (obelisk) [inline _src/doc-utils.ts:271]: Nit. Pull to top

File: _src/doc-utils.ts [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+  const switchToNetworkText = 'Switch network and add to wallet';
```
Comment (obelisk) [inline _src/doc-utils.ts:272]: Nit. Pull to top

File: _src/doc-utils.ts (hunk 3 above)
Comment (aelmanaa) [reply _src/doc-utils.ts:199]: @obelisk 1st of all thanks for taking the time to review. Actually it does. check the rest of the function:

- 1st it validates the format of an ethereum address by calling `isAddressFormatValid`
- then it connects it loads the contract object `new ethers.Contract` on the right network 
- finally it checks that the symbol contains LINK  and the number of decimals is correct

the goal of this function is to mitigate the risk of typo

File: _src/doc-utils.ts (hunk 2 above)
Comment (aelmanaa) [reply _src/doc-utils.ts:214]: ok @obelisk will remove this code in the next push

```
/*
  if (!equalOrContain(linkToken.name, name)) {
    // found inconsistencies on some networks . E.g: xdai, HECO chain..Etc
    throw new Error(`Error linkContract name. '${linkToken.name}' !== '${name}'`);
  }
  */
```

File: _src/doc-utils.ts (hunk 4 above)
Comment (aelmanaa) [reply _src/doc-utils.ts:267]: ok solved in next push

File: _src/doc-utils.ts (hunk 6 above)
Comment (aelmanaa) [reply _src/doc-utils.ts:271]: ok solved in next push

File: _src/doc-utils.ts (hunk 7 above)
Comment (aelmanaa) [reply _src/doc-utils.ts:272]: ok solved in next push

File: _src/doc-utils.ts (hunk 3 above)
Comment (obelisk) [reply _src/doc-utils.ts:187]: My mistake, apologies.

**Comment by obelisk** [comment]
This looks good to me now :) Thanks!

File: _src/doc-utils.ts [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+import chains from './reference/chains.json';
+import linkNameSymbol from './reference/linkNameSymbol.json';
+
+// disable unnecessary warnings
+ethers.utils.Logger.setLogLevel(ethers.utils.Logger.levels.ERROR);
+
+const chainlinkLogo =
+  'https://assets-global.website-files.com/5f6b7190899f41fb70882d08/5fa973e31adbc3414aa8f77d_Chainlink-webclip.png';
+
+const seperator = '_';
```
Comment (StephenFluin) [inline _src/doc-utils.ts:15]: Typo, should be `separator`

File: _src/index.ts [hunk 9]
```diff
@@ -8,3 +7,4 @@ import './notification.ts';
 import './subscribe.ts';
 import './scroll.ts';
 import './deploy-to-github.ts';
+import './doc-utils';
```
Comment (StephenFluin) [inline _src/index.ts:10]: Do we need t load this code on every page of the documentation? Is it worth the bundle size increase (and performance decrease), or should we only load it on specific pages?

File: _src/reference/chains.json [hunk 10]
```diff
@@ -0,0 +1,388 @@
+[
+  {
+    "name": "Ethereum Mainnet",
+    "chain": "ETH",
+    "icon": "ethereum",
+    "rpc": [
+      "https://mainnet.infura.io/v3/${INFURA_API_KEY}",
```
Comment (StephenFluin) [inline _src/reference/chains.json:7]: How are we going to maintain this file? These things change over time. Do we need all of this information for our simple "add to wallet" feature? it feels like there may be unnecessary fields here.

File: docs/Using Any API/contract-addresses.md [hunk 11]
```diff
@@ -20,7 +20,7 @@ For a comprehensive list of data providers, oracles, jobs, adapters, and more, v
 >
 > We strongly recommend that you deploy to a [testnet](#testnets) first, then move to mainnet.
 
-LINK address:  [`0x514910771af9ca656af840dff83e8264ecf986ca`](https://etherscan.io/token/0x514910771af9ca656af840dff83e8264ecf986ca)
+LINK address: <a class="erc-token-address" id="1_0x514910771af9ca656af840dff83e8264ecf986ca" href="https://etherscan.io/token/0x514910771af9ca656af840dff83e8264ecf986ca" target="_blank" rel="noreferrer, noopener">`0x514910771af9ca656af840dff83e8264ecf986ca`</a>
```
Comment (StephenFluin) [inline docs/Using Any API/contract-addresses.md:23]: We already add noreferrer noopener by JS I believe. Can you double check if we need this? I don't think we do.

**Review by StephenFluin — CHANGES_REQUESTED** [review]

File: docs/Using Any API/contract-addresses.md (hunk 11 above)
Comment (aelmanaa) [reply docs/Using Any API/contract-addresses.md:23]: @StephenFluin  we don't. For instance, open "_site/docs/deploy-your-first-contract/index.html" and you won't find noreferrer noopener on external links. We can look for a Eleventy plugin that does that by default for external links

File: _src/reference/chains.json (hunk 10 above)
Comment (aelmanaa) [reply _src/reference/chains.json:7]: @StephenFluin 
"add to wallet" feature has become "Switch to network (network must be added to metamask if not exist) then add to wallet". This data is required to add a network to metamask.
Originally, we were automatically pulling the info from this [source](https://chainid.network/chains.json) that is maintained by [chainlist](https://github.com/ethereum-lists/chains). Security requested to get a copy and validate it manually. Hence, every time that we support a new EVM chain, we pull data relevant to that chain, validate it and add it to "chains.json"

File: _src/index.ts (hunk 9 above)
Comment (aelmanaa) [reply _src/index.ts]: @StephenFluin resolved

File: docs/Using Any API/contract-addresses.md (hunk 11 above)
Comment (aelmanaa) [reply docs/Using Any API/contract-addresses.md]: resolved

**Review by dwightjl — APPROVED** [review]

**Comment by dwightjl** [comment]
Squashed and rebased to main

**Pull request merged**